src/commands/dev/socket.rs

```diff
@@ -0,0 +1,80 @@
+use std::time::Duration;
+
+use chrome_devtools as protocol;
+
+use console::style;
+
+use futures::{future, pin_mut, StreamExt};
+use futures_util::sink::SinkExt;
+
+use tokio::time;
+use tokio_tungstenite::connect_async;
+use tungstenite::protocol::Message;
+
+use url::Url;
+
+const KEEP_ALIVE_INTERVAL: u64 = 10;
+
+pub async fn listen(session_id: String) -> Result<(), failure::Error> {
+    let socket_url = format!("wss://rawhttp.cloudflareworkers.com/inspect/{}", session_id);
+    let socket_url = Url::parse(&socket_url)?;
+
+    let (ws_stream, _) = connect_async(socket_url)
+        .await
+        .expect("Failed to connect to devtools instance");
+
+    let (mut write, read) = ws_stream.split();
+
+    let enable_runtime = protocol::runtime::SendMethod::Enable(1.into());
+    let enable_runtime = serde_json::to_string(&enable_runtime)?;
+    let enable_runtime = Message::Text(enable_runtime.into());
+    write.send(enable_runtime).await?;
+
+    let (keep_alive_tx, keep_alive_rx) = futures::channel::mpsc::unbounded();
+    tokio::spawn(keep_alive(keep_alive_tx));
+    let keep_alive_to_ws = keep_alive_rx.map(Ok).forward(write);
+
+    let print_ws_messages = {
+        read.for_each(|message| async {
+            let message = message.unwrap().into_text().unwrap();
+            log::info!("{}", message);
+            let message: Result<protocol::Runtime, failure::Error> = serde_json::from_str(&message)
+                .map_err(|e| failure::format_err!("this event could not be parsed:\n{}", e));
+            if let Ok(protocol::Runtime::Event(event)) = message {
+                match event {
+                    protocol::runtime::Event::ConsoleAPICalled(event) => {
+                        match event.r#type.as_str() {
+                            "log" => println!("{}", style(event).green()),
+                            "error" => println!("{}", style(event).red()),
+                            _ => println!("{}", style(event).yellow()),
+                        }
+                    }
+                    protocol::runtime::Event::ExceptionThrown(event) => {
+                        println!("{}", style(event).bold().red())
+                    }
+                    _ => (),
+                }
+            }
+        })
+    };
+    pin_mut!(keep_alive_to_ws, print_ws_messages);
+    future::select(keep_alive_to_ws, print_ws_messages).await;
+    Ok(())
+}
+
+async fn keep_alive(tx: futures::channel::mpsc::UnboundedSender<Message>) {
```

Comment:
might be worth giving this a return type of `!`

Reply:
Interesting! I'm not familiar with this syntax and can't seem to find an example of a function returning `!` - would you mind elaborating?

Reply:
https://doc.rust-lang.org/stable/book/ch19-04-advanced-types.html#the-never-type-that-never-returns

Reply:
(`loop` has the type `!`, but since `!` coerces to any other type, it coerces to the return type of `()` here, which is why this code works the way it does)

Reply:
this is very cool